rust/src/detect/xbits.rs

```diff
@@ -0,0 +1,224 @@
+use std::ffi::{CStr, CString};
+use std::os::raw::{c_char, c_void};
+
+pub const DETECT_XBITS_CMD_SET: u8 = 0;
+pub const DETECT_XBITS_CMD_TOGGLE: u8 = 1;
+pub const DETECT_XBITS_CMD_UNSET: u8 = 2;
+pub const DETECT_XBITS_CMD_ISNOTSET: u8 = 3;
+pub const DETECT_XBITS_CMD_ISSET: u8 = 4;
+pub const DETECT_XBITS_CMD_NOALERT: u8 = 5;
+pub const DETECT_XBITS_CMD_MAX: u8 = 6;
+pub const DETECT_XBITS_TRACK_IPSRC: u8 = 0;
+pub const DETECT_XBITS_TRACK_IPDST: u8 = 1;
+pub const DETECT_XBITS_TRACK_IPPAIR: u8 = 2;
+pub const DETECT_XBITS_TRACK_FLOW: u8 = 3;
+pub const DETECT_XBITS_EXPIRE_DEFAULT: u32 = 30;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct RSXBitsData {
+    cmd: u8,
+    tracker: u8,
+    expire: u32,
+    vartype: u8,
+    name: *const c_char,
+}
+
+fn get_xbit_type(str_t: &str) -> Result<u8, ()> {
+    let res = match str_t {
+        "ip_src" => DETECT_XBITS_TRACK_IPSRC,
+        "ip_dst" => DETECT_XBITS_TRACK_IPDST,
+        "ip_pair" => DETECT_XBITS_TRACK_IPPAIR,
+        "flow" => DETECT_XBITS_TRACK_FLOW,
+        _ => return Err(()),
+    };
+    Ok(res)
+}
+
+fn get_xbit_cmd(str_c: &str) -> Result<u8, ()> {
+    let res = match str_c {
+        "set" => DETECT_XBITS_CMD_SET,
+        "toggle" => DETECT_XBITS_CMD_TOGGLE,
+        "unset" => DETECT_XBITS_CMD_UNSET,
+        "isnotset" => DETECT_XBITS_CMD_ISNOTSET,
+        "isset" => DETECT_XBITS_CMD_ISSET,
+        "noalert" => DETECT_XBITS_CMD_NOALERT,
+        _ => return Err(()),
+    };
+    Ok(res)
+}
+
+fn evaluate_args(args: Vec<&str>) -> Result<(u8, CString, u8, u32, u8), ()> {
+    let tracker: Vec<&str> = args[2].trim().split(' ').collect();
+    if tracker.len() != 2 {
+        return Err(());
+    }
+    match tracker[0].trim() {
+        "track" => {}
+        _ => {
+            SCLogError!("xbits track keyword parsing failed");
+            return Err(());
+        }
+    }
+    let tracker: u8 = match get_xbit_type(tracker[1].trim()) {
+        Ok(val) => val,
+        Err(e) => {
+            SCLogError!("xbits tracker parsing failed: {:?}", e);
+            return Err(());
+        }
+    };
+    let vartype = if tracker == DETECT_XBITS_TRACK_IPPAIR {
+        11 // VAR_TYPE_IPPAIR_BIT
```

Comment:
Is there a way to bring C constants to Rust without getter functions? Should we have a generic getter function for bringing constants from C that are not going to be brought here soon?

Reply:
Not at this time.  That would require `bindgen` which we don't have integrated at this time.  Best to just duplicate them as constants somewhere.